### ISSUE TITLE
legacy python provider is slow

### ISSUE THREAD

**Issue by tcbbd** [neovim/neovim#7063]
- `nvim --version`: NVIM v0.2.1-dev (git201707221704+3540+26~ubuntu16.04)
- Vim (version: vim 7.4 from Ubuntu official repo) behaves differently.
- Operating system/version: Ubuntu 16.04
- Terminal name/version: Gnome Terminal
- `$TERM`: xterm-256color

### Problem
The legacy python provider seems slow, plugins that use the legacy python interface can become laggy in neovim.  I think #2437 and #6843 is related.

### Steps to reproduce
Use [cpsm](https://github.com/nixprime/cpsm) matcher with ctrlp to speed up it and get better fuzzy matching. The problem is that cpsm is not performing well in neovim because it uses legacy python interface.

### Actual behaviour
In neovim, ctrlp is not very responsive in large projects (e.g. Linux kernel source), although acceptable to some degree. As an extreme example, if I invoke ctrlp in my home directory, when I type a letter, I must wait ~1s for the current matching results, then I can proceed to type the next letter.

### Expected behaviour
In vim, the matched results are shown instantly, and I can keep typing without waiting, even if I invoke it in my home directory. This smooth experience is expected.

### Profiling results
The difference is due to the performance of neovim legacy python provider.  As an example, let's try to search "memory.c" in linux kernel source.

neovim:
```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    9   1.894581   0.057493  <SNR>185_BuildPrompt()
    9   1.798171   0.000596  <SNR>185_Update()
    9   1.744472   0.000930  <SNR>185_MatchedItems()
    9   1.736736             cpsm#CtrlPMatch()
   10   1.731022             provider#python3#Call()
    8   1.636100   0.000152  <SNR>185_PrtFocusMap()
    8   1.635948   0.000261  <SNR>185_PrtAdd()
    1   0.448524   0.000339  ctrlp#init()
```

vim:
```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    9   0.446616   0.067764  <SNR>160_BuildPrompt()
    8   0.409632   0.000187  <SNR>160_PrtFocusMap()
    8   0.409445   0.000276  <SNR>160_PrtAdd()
    9   0.325209   0.000739  <SNR>160_Update()
    1   0.284742   0.000444  ctrlp#init()
    9   0.259212   0.001105  <SNR>160_MatchedItems()
    9   0.247648             cpsm#CtrlPMatch()
```

Obviously, `provider#python3#Call()` makes neovim slow. But I'm not sure why, maybe the RPC is slow?
```
FUNCTION  provider#python3#Call()
Called 10 times
Total time:   1.731022
 Self time:   1.731022

count  total (s)   self (s)
   10              0.000020   if s:err != ''
                                return
                              endif
   10              0.000020   if !exists('s:host')
                                let s:rpcrequest = function('rpcrequest')
                            
                                " Ensure that we can load the Python3 host before bootstrapping
                                try
                                  let s:host = remote#host#Require('legacy-python3-provider')
                                catch
                                  let s:err = v:exception
                                  echohl WarningMsg
                                  echomsg v:exception
                                  echohl None
                                  return
                                endtry
                              endif
   10              1.730780   return call(s:rpcrequest, insert(insert(a:args, 'python_'.a:method), s:host))
```

**Comment by tcbbd** [comment]
Strange! Even if I don't use python matcher for ctrlp (i.e. plain ctrlp), neovim is still slower than vim. That means vimL in neovim is slower!

vim:
```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    9   2.228333   0.021612  <SNR>160_BuildPrompt()
    9   2.172816   0.056648  <SNR>160_Update()
    8   2.165784   0.000161  <SNR>160_PrtFocusMap()
    8   2.165623   0.000197  <SNR>160_PrtAdd()
    9   1.072790   0.018499  <SNR>160_Render()
    9   1.041732   0.043547  <SNR>160_MatchedItems()
 4368   1.015956   0.212832  <SNR>160_mixedsort()
    9   0.997857   0.954010  <SNR>160_MatchIt()
17121   0.788090   0.563372  <SNR>160_matchlens()
 4368   0.674823   0.063662  <SNR>160_compmatlen()
    1   0.269805   0.000485  ctrlp#init()
    1   0.186684   0.000151  ctrlp#setlines()
    1   0.185977   0.000299  ctrlp#files()
    1   0.138748   0.010278  <SNR>160_UserCmd()
 5282   0.070716   0.015113  <SNR>160_curtype()
 5283   0.055619             <SNR>160_CurTypeName()
    9   0.043621             <SNR>160_sublist()
 4368   0.041661             <SNR>160_multipliers()
    2   0.040999   0.040884  ctrlp#utils#writecache()
    1   0.040583   0.000040  <SNR>160_writecache()
```

```
FUNCTION  <SNR>160_mixedsort()
Called 4368 times
Total time:   1.015956
 Self time:   0.212832

count  total (s)   self (s)
 4368   0.071458   0.011473 	let ct = s:curtype()
 4368              0.004495 	if ct == 'buf'
                            		let pat = '[\/]\?\[\d\+\*No Name\]$'
                            		if a:1 =~# pat && a:2 =~# pat | retu 0
                            		elsei a:1 =~# pat | retu 1
                            		elsei a:2 =~# pat | retu -1 | en
                            	en
 4368   0.733539   0.032061 	let [cln, cml] = [ctrlp#complen(a:1, a:2), s:compmatlen(a:1, a:2)]
 4368              0.003846 	if s:ispath
 4368              0.004217 		let ms = []
 4368              0.004711 		if s:res_count < 21
                            			let ms += [s:compfnlen(a:1, a:2)]
                            			if ct !~ '^\(buf\|mru\)$' | let ms += [s:comptime(a:1, a:2)] | en
                            			if !s:itemtype | let ms += [s:comparent(a:1, a:2)] | en
                            		en
 4368              0.020062 		if ct =~ '^\(buf\|mru\)$'
                            			let ms += [s:compmref(a:1, a:2)]
                            			let cln = cml ? cln : 0
                            		en
 4368              0.010216 		let ms += [cml, 0, 0, 0]
 4368   0.064308   0.022647 		let mp = call('s:multipliers', ms[:3])
 4368              0.018177 		retu cln + ms[0] * mp[0] + ms[1] * mp[1] + ms[2] * mp[2] + ms[3] * mp[3]
                            	en
                            	retu cln + cml * 2
```

neovim:
```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    9   2.646082   0.031927  <SNR>185_BuildPrompt()
    9   2.577766   0.034693  <SNR>185_Update()
    8   2.533670   0.000181  <SNR>185_PrtFocusMap()
    8   2.533489   0.000228  <SNR>185_PrtAdd()
    9   1.312411   0.064239  <SNR>185_MatchedItems()
    9   1.247812   1.171612  <SNR>185_MatchIt()
    9   1.228205   0.019906  <SNR>185_Render()
 4368   1.162887   0.243004  <SNR>185_mixedsort()
17121   0.905631   0.645269  <SNR>185_matchlens()
 4368   0.775627   0.074807  <SNR>185_compmatlen()
    1   0.162985   0.000429  ctrlp#init()
 5282   0.077170   0.018139  <SNR>185_curtype()
    9   0.075850             <SNR>185_sublist()
 5283   0.059053             <SNR>185_CurTypeName()
 4368   0.048495             <SNR>185_multipliers()
  859   0.041704   0.027195  <SNR>185_formatline()
 8736   0.039262             <SNR>185_shortest()
   11   0.038855   0.006456  325()
    9   0.035489   0.000849  ctrlp#statusline()
    9   0.034584   0.000426  airline#extensions#ctrlp#ctrlp_airline()
```

```
FUNCTION  <SNR>185_mixedsort()
Called 4368 times
Total time:   1.162887
 Self time:   0.243004

count  total (s)   self (s)
 4368   0.078725   0.013843 	let ct = s:curtype()
 4368              0.005216 	if ct == 'buf'
                            		let pat = '[\/]\?\[\d\+\*No Name\]$'
                            		if a:1 =~# pat && a:2 =~# pat | retu 0
                            		elsei a:1 =~# pat | retu 1
                            		elsei a:2 =~# pat | retu -1 | en
                            	en
 4368   0.845141   0.038634 	let [cln, cml] = [ctrlp#complen(a:1, a:2), s:compmatlen(a:1, a:2)]
 4368              0.004046 	if s:ispath
 4368              0.004991 		let ms = []
 4368              0.004858 		if s:res_count < 21
                            			let ms += [s:compfnlen(a:1, a:2)]
                            			if ct !~ '^\(buf\|mru\)$' | let ms += [s:comptime(a:1, a:2)] | en
                            			if !s:itemtype | let ms += [s:comparent(a:1, a:2)] | en
                            		en
 4368              0.019115 		if ct =~ '^\(buf\|mru\)$'
                            			let ms += [s:compmref(a:1, a:2)]
                            			let cln = cml ? cln : 0
                            		en
 4368              0.010792 		let ms += [cml, 0, 0, 0]
 4368   0.074472   0.025978 		let mp = call('s:multipliers', ms[:3])
 4368              0.018336 		retu cln + ms[0] * mp[0] + ms[1] * mp[1] + ms[2] * mp[2] + ms[3] * mp[3]
                            	en
                            	retu cln + cml * 2
```

**Comment by ZyX-I** [comment]
@tcbbd There *is* RPC, in Vim there is *no*. This makes a huge difference in how plugins need to be optimized to work faster.

About difference: I am not sure why VimL is slower in your test, but I see that airline calls do not appear in Vim output, meaning that there is some difference in setup. Also can you show full `nvim --version` output, as well as `vim --version`? Compiler and compilation flags are significant in such comparisons, as well as data like whether your Neovim has logging enabled, and at which level.

**Comment by justinmk** [comment]
#2437 and #6843 involve startup time (of nvim) only, not general performance.

> Obviously, provider#python3#Call() makes neovim slow.

No. Every python call goes through `provider#python3#Call()`, so that doesn't tell us anything.

Sorting on number of calls (instead of total time) will be more useful. Plugins that make numerous calls that are marshalled between VimL and python (e.g. iterating over a big VimL dict from python instead of pulling it over to python in one batch, then iterating) are known to be slower. That's just a fact. 

> That means vimL in neovim is slower!

I only see a ~10% difference, which could be related to many factors: 

- vim8 features
- more extensive validation of VimL objects in Nvim (better error messages)
- reading/writing files to/from disk may be slightly slower (but more reliable) in Nvim

**Comment by ZyX-I** [comment]
@justinmk What validation are you talking about? About “better error messages” I know for sure only about my part, that is implementation of JSON (which still has no error messages more useful then “invalid argument”), viminfo replacement and some places where error messages were changed to include `os_strerror()` output.

**Comment by tcbbd** [comment]
> Also can you show full nvim --version output, as well as vim --version?

`nvim --version`:
```
NVIM v0.2.1-dev
Build type: RelWithDebInfo
Compilation: /usr/bin/x86_64-linux-gnu-gcc -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wconversion -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_MSGPACK_HAS_FLOAT32 -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/build/config -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/src -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/.deps/usr/include -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/.deps/usr/include -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/.deps/usr/include -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/.deps/usr/include -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/.deps/usr/include -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/.deps/usr/include -I/usr/include -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/build/src/nvim/auto -I/build/neovim-iaPgIb/neovim-0.2.0ubuntu1+git201707221704+3540+26~ubuntu16.04.1/build/include
Compiled by root@lcy01-09

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui      
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
```

`vim --version`:
```
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Nov 24 2016 16:44:48)
包含补丁: 1-1689
另外的补丁： 8.0.0056
修改者 pkg-vim-maintainers@lists.alioth.debian.org
编译者 pkg-vim-maintainers@lists.alioth.debian.org
巨型版本 with GTK3 GUI.  可使用(+)与不可使用(-)的功能:
+acl             +farsi           +mouse_netterm   +tag_binary
+arabic          +file_in_path    +mouse_sgr       +tag_old_static
+autocmd         +find_in_path    -mouse_sysmouse  -tag_any_white
+balloon_eval    +float           +mouse_urxvt     +tcl
+browse          +folding         +mouse_xterm     +terminfo
++builtin_terms  -footer          +multi_byte      +termresponse
+byte_offset     +fork()          +multi_lang      +textobjects
+channel         +gettext         -mzscheme        +timers
+cindent         -hangul_input    +netbeans_intg   +title
+clientserver    +iconv           +packages        +toolbar
+clipboard       +insert_expand   +path_extra      +user_commands
+cmdline_compl   +job             +perl            +vertsplit
+cmdline_hist    +jumplist        +persistent_undo +virtualedit
+cmdline_info    +keymap          +postscript      +visual
+comments        +langmap         +printer         +visualextra
+conceal         +libcall         +profile         +viminfo
+cryptv          +linebreak       -python          +vreplace
+cscope          +lispindent      +python3         +wildignore
+cursorbind      +listcmds        +quickfix        +wildmenu
+cursorshape     +localmap        +reltime         +windows
+dialog_con_gui  +lua             +rightleft       +writebackup
+diff            +menu            +ruby            +X11
+digraphs        +mksession       +scrollbind      -xfontset
+dnd             +modify_fname    +signs           +xim
-ebcdic          +mouse           +smartindent     +xsmp_interact
+emacs_tags      +mouseshape      +startuptime     +xterm_clipboard
+eval            +mouse_dec       +statusline      -xterm_save
+ex_extra        +mouse_gpm       -sun_workshop    +xpm
+extra_search    -mouse_jsbterm   +syntax          
     系统 vimrc 文件: "$VIM/vimrc"
     用户 vimrc 文件: "$HOME/.vimrc"
 第二用户 vimrc 文件: "~/.vim/vimrc"
      用户 exrc 文件: "$HOME/.exrc"
    系统 gvimrc 文件: "$VIM/gvimrc"
    用户 gvimrc 文件: "$HOME/.gvimrc"
第二用户 gvimrc 文件: "~/.vim/gvimrc"
        系统菜单文件: "$VIMRUNTIME/menu.vim"
         $VIM 预设值: "/usr/share/vim"
编译方式: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-3.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib/x86_64-linux-gnu/dbus-1.0/include -I/usr/include/gtk-3.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/mirclient -I/usr/include/mircommon -I/usr/include/mircookie -I/usr/include/cairo -I/usr/include/pango-1.0 -I/usr/include/harfbuzz -I/usr/include/pango-1.0 -I/usr/include/atk-1.0 -I/usr/include/cairo -I/usr/include/pixman-1 -I/usr/include/freetype2 -I/usr/include/libpng12 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libpng12 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -Wdate-time  -g -O2 -fPIE -fstack-protector-strong -Wformat -Werror=format-security -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1      
链接方式: gcc   -L. -Wl,-Bsymbolic-functions -Wl,-z,relro -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E  -Wl,-Bsymbolic-functions -fPIE -pie -Wl,-z,relro -Wl,-z,now -Wl,--as-needed -o vim   -lgtk-3 -lgdk-3 -lpangocairo-1.0 -lpango-1.0 -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lSM -lICE -lXpm -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux  -lacl -lattr -lgpm -ldl  -L/usr/lib -llua5.2 -Wl,-E  -fstack-protector-strong -L/usr/local/lib  -L/usr/lib/x86_64-linux-gnu/perl/5.22/CORE -lperl -ldl -lm -lpthread -lcrypt  -L/usr/lib/python3.5/config-3.5m-x86_64-linux-gnu -lpython3.5m -lpthread -ldl -lutil -lm -L/usr/lib/x86_64-linux-gnu -ltcl8.6 -ldl -lz -lpthread -lieee -lm -lruby-2.3 -lpthread -lgmp -ldl -lcrypt -lm     
```

**Comment by justinmk** [comment]
@ZyX-I The typval traversal is implemented differently in Nvim. Also `string()` output differs, IIRC there were some performance implications with how dicts are traversed. And I thought you added more granular error messages for eval.

**Comment by ZyX-I** [comment]
Neovim has `-Og` as last optimization flag as far as I see. Vim has `-O2`.

**Comment by tcbbd** [comment]
> but I see that airline calls do not appear in Vim output, meaning that there is some difference in setup.
Airline is in use in both vim and neovim, I use the same vimrc for vim & neovim (except there's some options & plugins only activated in neovim, as they only work in neovim)

**Comment by tcbbd** [comment]
@ZyX-I The compilation flag is a good explanation, thanks

**Comment by ZyX-I** [comment]
@justinmk More granular eval error messages are in that unfinished VimL to lua translator PR which does not  look like actually going to be finished. In master they should appear as one of the next steps after #6364.

**Comment by ZyX-I** [comment]
Typval traversion is different, yes. But I do not remember adding much of error messages there, aside from more verbose errors when dumping recursive containers.

**Comment by Shougo** [comment]
@tcbbd Have you build neovim as Release version?

https://github.com/neovim/neovim/wiki/Building-Neovim#optimized-builds

**Comment by tcbbd** [comment]
> Sorting on number of calls (instead of total time) will be more useful. Plugins that make numerous calls that are marshalled between VimL and python (e.g. iterating over a big VimL dict from python instead of pulling it over to python in one batch, then iterating) are known to be slower. That's just a fact. #1898 

I think there're not so much calls in my case. Only 10 calls to python. I suspect the large amount of time spent in `provider#python3#Call()` is due to data copying from neovim to python. Here is the `cpsm.py` script, it just get the file list of ctrlp from vim and then feed it into a compiled module.
```python
def ctrlp_match():
    try:
        results, regexes = cpsm_py.ctrlp_match(
                vim.eval("a:items"), vim.eval("a:str"),
                limit=int(vim.eval("a:limit")), mmode=vim.eval("a:mmode"),
                ispath=int(vim.eval("a:ispath")), crfile=vim.eval("a:crfile"),
                highlight_mode=vim.eval("g:cpsm_highlight_mode"),
                match_crfile=int(vim.eval("s:match_crfile")),
                max_threads=int(vim.eval("g:cpsm_max_threads")),
                query_inverting_delimiter=vim.eval("g:cpsm_query_inverting_delimiter"),
                regex_line_prefix=vim.eval("s:regex_line_prefix"),
                unicode=int(vim.eval("g:cpsm_unicode")))
        # Escape backslashes and ".
        vim.command("let s:results = [%s]" % ",".join(
                map(_escape_and_quote, results)))
        vim.command("let s:regexes = [%s]" % ",".join(
                map(_escape_and_quote, regexes)))
    except Exception as ex:
        ex_str = str(ex)
        if "function takes at most" in ex_str:
            # Most likely due to a new parameter being added to
            # cpsm_py.ctrlp_match.
            ex_str = "rebuild cpsm by running %s: %s" % (
                    os.path.normpath(os.path.join(
                            script_dir, "..", "install.sh")),
                    ex_str)
        vim.command("let s:results = [%s]" % _escape_and_quote(
                "ERROR: " + ex_str))
        vim.command("let s:regexes = []")
```

**Comment by tcbbd** [comment]
@Shougo I get it from [ppa](https://launchpad.net/%7Eneovim-ppa/+archive/ubuntu/unstable)

**Comment by wsdjeg** [comment]
As I know ctrlp run different vim script in vim and neovim, It only use job in vim  8. And use system and system list in neovim and old vim. When use job . The matcher do not need to filters all files.

**Comment by Shougo** [comment]
Related issue: https://github.com/neovim/neovim/issues/5702

**Comment by tcbbd** [comment]
@Shougo @ZyX-I Tested with optimized neovim build, now neovim is faster!
```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    9   2.086580   0.034588  <SNR>185_BuildPrompt()
    8   2.024475   0.000213  <SNR>185_PrtFocusMap()
    8   2.024262   0.000309  <SNR>185_PrtAdd()
    9   2.019891   0.037544  <SNR>185_Update()
    9   0.996904   0.015088  <SNR>185_Render()
    9   0.982764   0.045368  <SNR>185_MatchedItems()
 4368   0.947474   0.198323  <SNR>185_mixedsort()
    9   0.936969   0.899135  <SNR>185_MatchIt()
17121   0.744025   0.528936  <SNR>185_matchlens()
 4368   0.631073   0.056433  <SNR>185_compmatlen()
    1   0.207125   0.000190  ctrlp#init()
    1   0.133168   0.000047  ctrlp#setlines()
    1   0.133028   0.000205  ctrlp#files()
    1   0.116263             <SNR>185_UserCmd()
 5282   0.061823   0.012973  <SNR>185_curtype()
 5283   0.048858             <SNR>185_CurTypeName()
 4368   0.040560             <SNR>185_multipliers()
    9   0.037553             <SNR>185_sublist()
   18   0.036491   0.002311  airline#check_mode()
   11   0.036126   0.006039  325()
```

But the python provider performance doesn't change :(

**Comment by ZyX-I** [comment]
@tcbbd It should not. Bottleneck is IPC in this case, not lack of optimization: every `:python`/`pyeval()` is something like.

1. Generate request. Calls some VimL functions from `provider#*`.
2. Serialize request. C, uses first internal API types which are then serialized via msgpack-c library.
3. Send request. Includes switching to kernel. May need multiple transactions.
4. (Python) Receive request. Includes switching from kernel. May need multiple transactions.
5. (Python) Deserialize request.
6. (Python) Process request.
6.1. Includes “generate -> serialize -> send -> receive -> deserialize -> process -> serialize -> send -> receive -> deserialize -> process” each time Neovim API is being used, though in the opposite direction (first generate on Python client).
7. (Python) Serialize response.
8. (Python) Send response. Includes switching to kernel. May need multiple transactions.
9. Receive response. Includes switching from kernel. May need multiple transactions.
10. Deserialize response. C, uses same msgpack-c library, but now it is always “first create msgpack-c objects, then convert them to internal API objects, then free msgpack-c”.
11. Process response. Should be the fastest step here.

In Vim serialize/send/receive/deserialize are folded into a single step “convert internal object (e.g. typval_T) into Python object”. In some cases this includes copying everything around with encoding conversion, in other it is just a mere “create a Python object, save pointer to Vim object inside, save information that Vim object is referenced in it”. But in any case Vim not only omits that serializing/deserializing mess, but also switching between kernel and user mode. Also Vim does not need to run additional VimL code to communicate with Python.

Our approach has many advantages, but performance is not among them.

**Comment by tcbbd** [comment]
@ZyX-I So vim is faster with cpsm.py because it embeds a python in it so the large file list is directly accessed by python, but in neovim the file list should be sent to external python process via IPC. If I understand it correctly, this kind of performance issue (large amount of communication between neovim and python or any other language provider) is an inevitable result of our design. So we should modify/redesign plugins to mitigate the problem?

**Comment by Shougo** [comment]
> So we should modify/redesign plugins to mitigate the problem?

Yes, the modification(optimization) is needed for the plugin.
For example, UltiSnips is not optimized with neovim.

**Comment by bluddy** [comment]
@ZyX-I 
> More granular eval error messages are in that unfinished VimL to lua translator PR which does not look like actually going to be finished.

If I may ask, why is this the case? It was a pretty interesting part of the original Neovim design.

**Comment by ZyX-I** [comment]
@bluddy Basically two problems:

1. Original PR is too big for one PR, not ready for splitting and one can’t start making benefits of partial merging. One of the reasons #6364 exists is that I am going to use parser for highlighting and provide benefits for users without having anything fully finished. But even to use what I have there for highlighting I need rather big refactoring because I did not thought about that in advance.
2. When I started implementing actually translator and not just parser it felt like I am trying to fit a square into a round hole: lua has different semantics, so not only the result is only going to be fast with much effort, but it would be easy to leak lua semantics: e.g. can’t add string and number, but can concat string and float, yet can’t concat luajit integer and string; what is worse are things like depending on presence of `:try` execution may or may not continue after error.

I am now thinking of creating a VM instead, though not sure whether such creating would be “write our own from scratch” or “take existing one and refactor it”. Thought about Parrot, but it appears to be dead now (last commit is 9 Nov 2016) and I saw some articles critiquing its design. In any case in order to be incremental VM will only be created after parser is finished, though it is possible to start to create it after finishing only expressions parser since there are lambdas now (they may use it right there since they absolutely need only expressions parser).

**Comment by bluddy** [comment]
@ZyX-I to be honest, I was skeptical about the performance of implementing an interpreter inside an interpreter, but you seemed to have it under control. Perhaps this isn't worth pursuing too far -- if Neovim just adopts lua as the first-class scripting language (while keeping vimL support) and enhances the API for Neovim-specific stuff, that might be good enough.

**Comment by tcbbd** [comment]
@bluddy I think it's just about directly implementing a VimL VM. It is a good idea to improve VimL, but there seems to be great amount of work to be done and this plan would lie on the long-term todo list for a long time. After all, we have many other things to fix and to improve in Neovim now, and currently remote plugins work well (except for the pathological case of this issue, i.e. great amount of IPC communication).

**Comment by ZyX-I** [comment]
@bluddy I only fully implemented a parser, but even that had a bunch of bad design choices, unfinished translator implementation appeared to be slow for anything more or less complex. What I want now out of the parser is

1. If requested, create garray with highlighting alongside with the AST.
2. Both AST and highlighting need to be exposed via API. To be used in tests and in things like VimL linters, testing via a dumper is strange (though dumper is still good to have).
3. AST need to be more closely tighed up with source.

Lua is now going to be present always, and it has direct (i.e. via wrappers written in C which convert lua types to internal API types and directly call C functions) access to Neovim API. There is not Vim compatibility layer yet though, you may use only API.

@tcbbd The biggest problem with VimL is that it executes a string during parsing, but has not even AST, so strings need to be reparsed each time. I predict that even replacing “execute a string during parsing, do this each time function is called” with “parse to AST at function definition, execute AST ever since” as long as everything is C would be a big improvement for functions.

The IPC has nothing to do with the problem.

**Comment by bluddy** [comment]
@ZyX-I 
>  I predict that even replacing “execute a string during parsing, do this each time function is called” with “parse to AST at function definition, execute AST ever since” as long as everything is C would be a big improvement for functions.

This is a great idea. It probably makes more sense to do this than anything else. Probably the fastest approach is to use yacc and lex, with a C-based AST. Lua (even luajit) is just a layer of slowdown. Create a simple stack-based interpreter for the AST and you'll get pretty good performance.

BTW I realize it has nothing to do with this particular PR. I apologize -- I just saw your point and wanted to get more info.

**Comment by bluz71** [comment]
I independently came across the same issue as the original poster, I am a new *cpsm* user, and I also noted the Neovim slowdown. I opened (then closed) an [issue](https://github.com/nixprime/cpsm/issues/39) on the *cpsm* issue tracker.

The Neovim slowdown makes sense once RPC is explained.

I have a couple of questions:

* Why is the issue still open? Is there work proposed in Neovim to help with RPC heavy plugins? I can't see that can be done from the descriptions thus far.

* So the real fix will be needed in the *cpsm* plugin correct? If so what would that be? Replacing all the Python with Lua? Note, cpsm is a Python plugin that calls into C code. If it was converted/forked into Lua plugin that calls into C code it would be fast in Neovim? Or not?

* If a Lua version of *cpsm* is the solution then we need to wait since 0.2.0 still doesn't have the in-process Lua interpreter yet, or am I wrong?

If Lua is not the answer to speed up *cpsm* in Neovim then can some explain what is the best way to speed up *cpsm* in Neovim.

Note, *fzf* is not an answer or a solution for some of us. [I use CtrlP + ripgrep + cpsm](https://bluz71.github.io/2017/10/26/turbocharge-the-ctrlp-vim-plugin.html) since it never launches any external window when using a GUI based vim. I simply prefer a fast CtrlP over *fzf*; but I just wish *cpsm* was faster in Neovim than it currently is.

**Comment by justinmk** [comment]
> Why is the issue still open? Is there work proposed in Neovim to help with RPC heavy plugins

It's a bug, and there's still some potential for improvement.

> So the real fix will be needed in the cpsm plugin correct? If so what would that be? Replacing all the Python with Lua? 

No. The general idea is something like this:

- Look for places where a big list of items is being processed in a loop. The processing should be done in python, then the list sent to Nvim. That will solve the issue.
- If some decisions about each item are needed from Nvim, send the list as a batch to Nvim, work on it server-side (i.e. in Nvim), then return the result back to the python-side.
    - You could use Lua server-side to make things faster.

> If it was converted/forked into Lua plugin that calls into C code it would be fast in Neovim? Or not?

Yes, that works too. However, Lua runs in-process, so the "async" potential of a remote python plugin would be lost. (However, we will eventually ship a "remote host" for Lua as well.)

**Comment by bluz71** [comment]
@justinmk,

> It's a bug, and there's still some potential for improvement.

Ok, maybe some performance improvement in the future? If so, most welcome.

> Look for places where a big list of items is being processed in a loop....

I suspect big item processing is already occurring in C. The [cpsm](https://github.com/nixprime/cpsm) plugin is a custom matcher for the CtrlP plugin. I think the pipeline would be something like this:

User interaction in *CtrlP* (VimL) --- RPC ---> cpsm Python interface --- dynamic linked ---> cpsm C library which does all heavy lifting (matching and list processing). Don't quote me, but that is what I think would be happening. Hence the user pressing keys is in VimL that then gets passed via Vim's Python interface to a C-library that does the fuzzy matching and ranking.

Since Vim 8 with cpsm is super-duper fast (like nearly instant) whilst Neovim can lag (with huge source trees) it does seem the RPC interface (passing keypresses to the C-library and passing matches from the C-library back) is the issue. I don't think there is scope for *cpsm* to batch process since I think it would already do that (in C).

> Yes, that works too. However, Lua runs in-process, so the "async" potential of a remote python plugin would be lost. 

CtrlP is not async (as far as I know); hence not having async is no big deal. Which version of Neovim has Lua in-process? Am I right that Neovim 0.2.0 (the current stable version) does *not* have Lua built in? But 0.3.0 will have Lua?

I ask since I likely will create an issue at *cpsm* about whether a *cpsm* fork based on Lua for Neovim is possible. Bypassing RPC is likely the key for fast *cpsm* in Neovim from how I read this issue.

**Comment by justinmk** [comment]
> User interaction in CtrlP (VimL) --- RPC ---> cpsm Python interface --- dynamic linked ---> cpsm C library which does all heavy lifting (matching and list processing). Don't quote me, but that is what I think would be happening

That diagram is uni-directional, and doesn't tell us anything we didn't already know. Yes, RPC happens somewhere and does something. No, sending a list of files over RPC is not going to be slow. You can send _megabytes_ of data without a perceptible slowdown.

> Bypassing RPC is likely the key for fast cpsm in Neovim from how I read this issue

RPC is too vague here. Every Nvim GUI is implemented via RPC, and they are doing a lot more than sending a list of 100 files or whatever.

> 0.3.0 will have Lua?

0.2.1 will have Lua built-in. Pre-builds are available at https://github.com/neovim/neovim/releases .

**Comment by bluz71** [comment]
@justinmk,

I interpreted ZyX-I's response [here](https://github.com/neovim/neovim/issues/7063#issuecomment-317296032) as being the final answer. Since Neovim does Python plugins via RPC/IPC it can never be as fast as Vim 8 with the same plugins; which made sense.

How do you suggest we make *cpsm* as fast in Neovim as it is in Vim? Is a Neovim optimization needed or does the *cpsm* plugin need a Neovim enhancement?

Maybe the answer is cpsm never will be as fast, use a different fuzzy matching plugin?

**Comment by justinmk** [comment]
I addressed those questions above, in https://github.com/neovim/neovim/issues/7063#issuecomment-340227588 . Don't know what else to say.

**Comment by ZyX-I** [comment]
@bluz71 Just optimize that plugin with “communications with Neovim has big throughput *and* big latency” idea in mind while Vim has big throughput, small latency, but more likely to break when you use threads or fork-without-exec-based multiprocessing. Serialization/deserialization cycle with context switching do contribute to throughput as well, but its contribution to latency is far more noticeable.

This is basically what @justinmk suggested: in place of communicating in small chunks, do whatever is needed to merge them into bigger ones and process them in a batch on whatever side you need to process them: rather typical advice for Web services as well.

**Comment by bluz71** [comment]
@ZyX-I and @justinmk,

[cpsm](https://github.com/nixprime/cpsm) is *not* my plugin, @nixprime is the author and maintainer. I had a quick look at the code, but I don't understand how it does what it does (it is mostly C++). Your suggestions are excellent, but it may be the case that the plugin already does those optimisations. It is the latency that is the killer here, not throughput.

I will nonetheless open up an issue at the *cpsm* issue tracker noting and linking the suggestions here.

I've already mentioned, but my gut feel is that *cpsm* would need to be forked, say to *cpsm.nvim* with the existing Python bridge replaced by in-process Lua targetting Neovim 0.2.1 and onwards. That should have performance parity with *cpsm* running in Vim 8.0. I also expect this not to happen since @nixprime has been very quiet on Github for most of the year.

Lastly, I understand the benefits of the existing Neovim RPC/IPC architecture. The one penalty will be extreme latency dependent plugins; this particular issue may well be the worst case scenario for that latency penalty. Basically we live with it in Neovim, the plugin needs to change some way, some how (if the author has the will).

Thanks everyone. Good chat.

P.S. When fzf stops spawning a window in gVim maybe I'll replace CtrlP with *fzf*.

**Comment by ZyX-I** [comment]
@bluz71 I doubt having an issue *only* in cpsm will do any good, I have not checked CtrlP source code, but most likely you need to either make somebody modify only CtrlP or both CtrlP and cpsm: based on the README cpsm is on the receiving end, it being a plugin to a plugin means that CtrlP defines the API, and API definition is highly likely to be a major limiter unless cpsm is badly coded from the beginning.

**Comment by ZyX-I** [comment]
Though you need to check API defined by CtrlP first before doing such moves.

**Comment by bluz71** [comment]
@ZyX-I,

Vim 8 + CtrlP + cpsm has no latency at all; results are immediate (as fast as *fzf* with huge trees like Linux). It is surprising how responsive CtrlP + ripgrep + cpsm is even with huge source trees when in Vim 8.

With that data point I don't see why CtrlP needs to change to help fix this issue.

If Vim 8 is super fast (with CtrlP + cpsm) and Neovim has latency issues then RPC/IPC is most likely the issue? or no? Replace remote Python with in-process Lua should result in parity?

**Comment by ZyX-I** [comment]
CtrlP + cpsm may be fast even if e.g. cpsm has to call ctrlp#add_match() once for each and every match it found and check return value (example, I am not saying it does): that is in-process and fast in Vim. But if CtrlP does not have e.g. ctrlp#add_match_list() it will not be fast in Neovim with this approach, no matter how fast are CtrlP and cpsm individually.2:41, 30 октября 2017 г., Dennis B <notifications@github.com>:@ZyX-I,Vim 8 + CtrlP + cpsm has no latency at all; results are immediate (as fast as fzf with huge trees like Linux). It is surprising how responsive CtrlP + ripgrep + cpsm is even with huge source trees when in Vim 8.With that data point I don't see why CtrlP needs to change to help fix this issue.If Vim 8 is super fast (with CtrlP + cpsm) and Neovim has latency issues then RPC/IPC is most likely the issue? or no? Replace remote Python with in-process Lua should result in parity?—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by Shougo** [comment]
@bluz71 You can test denite.nvim instead.  It is optimized for neovim.  It also supports cpsm matcher.

**Comment by Shougo** [comment]
denite.nvim has not the issue.
Because denite.nvim is almost Python based plugin.
RPC calls are very few.

**Comment by bluz71** [comment]
@ZyX-I,

I did some digging using the *cpsm* matcher along with an alternate pure-Python matcher [ctrlp-py-matcher](https://github.com/FelikZ/ctrlp-py-matcher).

The pure Python matcher was easier to get a handle on.

CtrlP matchers need to provide a VimL function of the form:

```
function! Function_Name(items, str, limit, mmode, ispath, crfile, regex)
a:items  : The full list of items to search in.
a:str    : The string entered by the user.
...
return list_of_matched_items
```

It appears matches are already handled in bulk by the existing CtrlP API. The function returns a list of matched items given a set of *items* and *str*.

The realy latency killer in Neovim seems to by this single Python statement in a CtrlP matcher:

```
items = vim.eval('a:items')
```

[This](https://github.com/FelikZ/ctrlp-py-matcher/blob/master/autoload/pymatcher.py) is an example of a Python CtrlP matcher using the `vim.eval` (line 7). Comment that out and Neovim CtrlP is fast like Vim 8. The *cpsm* matcher also has the same `vim.eval`.

What is `vim.eval("a:items")` doing in Neovim when given an array of items (let's say 60,000 items)?

Lastly the latency issue really manifests when holding down the backspace key. In CtrlP I'll do `jjjjjjjjjj` then I'll hold down `<backspace>` then do `jjjjjjjjjj` again and do. Extremely unrealistic, but it highlights the latency slowdown. Vim 8 does not have the issue and Neovim will also not have any once `vim.eval('a:items')` is commented out.

Can Pythonic `vim.eval("a:items")` be made faster in Neovim than it currently is?

Interesting. Not what I expected.

**Comment by Shougo** [comment]
@bluz71 It is expected result.  `vim.eval()` is slow.  Especially the RPC data is huge.

**Comment by bluz71** [comment]
@Shougo,

`a:items` would have 60,000 strings (filename candidates) when using CtrlP on the Linux source tree.

Interesting that Neovim `vim.eval('a:items')` should be so slow. Nothing can be done?

Anyway, it seems that the CtrlP matchers are not doing the pathological case of populating the match list across a Python <-> VimL bridge one by one. Bulk processing appears to be happening. Rather it is the bulk action of getting the candidate list of possible filenames across the VimL <-> Python bridge that is a killer (maybe the killer here).

**Comment by Shougo** [comment]
Yes, Ctrlp-cpsm matcher implementation is not so bad.

> Interesting that Neovim vim.eval('a:items') should be so slow. Nothing can be done?

Yes, nothing.  For Ctrlp, using if_lua is better than if_python in neovim.

**Comment by bluz71** [comment]
Let's see what @ZyX-I says, but I tend to agree with you that the Python-in-ctrlp-matchers needs to be replaced completely if `vim.eval('a:items')` can not be made fast in Neovim.

Am I correct, Neovim 0.2.1 with in-built Lua runtime will be needed? Current stable Neovim 0.2.0 is not enough since in-process Lua does not yet exist?

Maybe I should drop *cpsm* and use *ctrlp-py-matcher* in Vim 8 and fork py-matcher into *ctrl-lua-matcher* for Neovim? I wonder how hard it is to port Python code to Lua?

**Comment by ZyX-I** [comment]
One of the problems with specifically `vim.eval` is that in *Vim* `vim.eval` implementation has a bad habit of coercing numbers to strings. But when you use `nvim_eval()` API function numbers are returned as numbers! This means that Python API implementation needs to receive that long list and go recursively through its items to convert numbers to strings itself. If that is the issue (and not that you are hitting a throughput problem actually instead) it should be mitigated by using native `neovim` API, `neovim.attach(…).eval(…)` does not do such conversions.

Additionally note that there is a number of `vim.eval` calls there in sequence which really can be merged all together.

**Comment by ZyX-I** [comment]
In any case you need to check profiling results first before speculating on “why `vim.eval('a:items')` is slow”.

**Comment by bluddy** [comment]
Is the python msgpack serialization written in python or C? If it's python, there's your bottleneck. Low-level serialization needs to be as fast as possible or it won't scale, and in python, that means calling out to a C implementation. If it is written in python, converting it to C will solve the problems in almost all plugins.

Also, it would be nice (but not absolutely necessary) of ctrl-p to have plugins cache their item list, so that it doesn't have to send the whole list of files with every command -- 90% of the time that list is't going to change.

**Comment by tcbbd** [comment]
@bluddy I think the deserialization (we are pulling an item from vim to python through `vim.eval('a:items')`) happens in python-msgpack implementation, the neovim python client just calls `msgpack.Unpacker.feed(data)` to get the results. So if the (de)serialization is slow, it means the python implementation of msgpack is slow.

**Comment by bluddy** [comment]
@tcbbd python-msgpack looks like it uses some C, or at least some Cython, but it's possible that it's losing some performance by making a python call somewhere, or that the neovim-client does something that slows the process down. Since this is a critical path for neovim, it needs to be optimized as heavily as possible and benchmarked.

It's one thing to have serialization/deserialization and message passing. I don't think it's a big deal as far as human perception is concerned. It's another thing to ever perform some invisible action on that data with a slow dynamic language like python -- especially something that isn't of constant computational complexity.

**Comment by ZyX-I** [comment]
@bluddy Are you sure that list in question is not already cached, but on the CtrlP side? It does not need to regenerate a list to pass the same reference again and again after all.

Also I see that Python’s msgpack is C entirely, msgpack/__init__.py just imports Packer, unpack, unpackb and Unpacker from two .so libraries and calls them (unless it is not available or MSGPACK_PUREPYTHON environment variable was set in which case pure Python fallback is used). But `vim.eval` still has to traverse the resulting structure to convert integers, and that is pure Python.

Some small benchmark regarding `LegacyVim.eval` vs `Nvim.eval`:

```
nvim -i NONE -u NONE -N --cmd 'command -nargs=1 TI let s=reltime() | execute <q-args> | echo reltimestr(reltime(s))' --cmd 'let longlist = map(range(100000), "0.v:val")' --cmd 'python import vim' --cmd 'TI python vim.eval("longlist")' --cmd 'python from neovim import Nvim' --cmd 'TI python Nvim.eval(vim, "longlist")' --cmd cq
  0.588690
  0.378885
```

. The results are pretty consistent, and are worse (as expected) if I remove `0.` (i.e. do not stringify values first in order to make `LegacyVim.eval` stringify them):

```
nvim -i NONE -u NONE -N --cmd 'command -nargs=1 TI let s=reltime() | execute <q-args> | echo reltimestr(reltime(s))' --cmd 'let longlist = map(range(100000), "v:val")' --cmd 'python import vim' --cmd 'TI python vim.eval("longlist")' --cmd 'python from neovim import Nvim' --cmd 'TI python Nvim.eval(vim, "longlist")' --cmd cq
  0.459307
  0.287371
```

: you see, LegacyVim.eval is 60% worse and basically the only difference is stringifying.

Note though that `nvim` itself in this test is a debug build, but not any of the dependencies (`neovim` python module, msgpack library used by Neovim, msgpack dependency used by Python, …).

**Comment by ZyX-I** [comment]
Though also note that 0.38 seconds for transferring 100 000 strings in a list with what is supposed to be almost pure C. Vim itself copes in 0.02 seconds with basically the same code (minus commands starting from `python from neovim …` for obvious reasons).

**Comment by bfredl** [comment]
FWIW with `import vim` the faster nvim eval can be accessed as `vim.api.eval`

**Comment by ZyX-I** [comment]
Release build shows mostly the same results: something like 0.42 vs 0.21, though results are less stable and I can’t say (no) pre-stringifying makes any difference. Now though it is not 60% worse, time was doubled.

**Comment by bluz71** [comment]
@ZyX-I,

> @bluddy Are you sure that list in question is not already cached, but on the CtrlP side? It does not need to regenerate a list to pass the same reference again and again after all.

CtrlP does have file caching which you can enable or disable. However, with external matchers it appears to always send the full candidate file list whether regenerated (if CtrlP caching is off) or cached. Matchers themselves don't cache the file list.

In an ideal world I disable CtrlP caching since it causing problems with new files or deleted files. The cache list always ends up stale in my experience.

**Comment by bluz71** [comment]
@ZyX-I,

I would like to test `Nvim.eval` in the *cpsm* Python matcher code.

Currently the code is something like:

```
import vim
items = vim.eval('a:items')
```

I converted it to (which does not work), what is my mistake?

```
import Nvim
items = Nvim.eval(vim, "a:items")
```

**Comment by bluz71** [comment]
@ZyX-I,

Fixed it, I needed `from neovim import Nvim`.

Quick results, `items = Nvim.eval(vim, "a:items")` is **MUCH** faster than `items = vim.eval('a:items')`.

Not as fast as Vim 8 but a huge improvement. Your quick comparison above noted a 2 times speed improvement. Yes, it feels twice as fast.

Currently I have tested on my fast Linux box; I will now test on my slow Macbook and report back.

**Comment by bluddy** [comment]
@ZyX-I, is the python iteration the thing that's dominating those 0.38 seconds? If so, is it possible to call out to C when iterating over those large data structures? I'd love to see the python backend compared to a pure C backend, to see how fast this can be made. Your (n)vim-foo is clearly (and unsurprisingly) far superior to mine.

**Comment by bluddy** [comment]
@ZyX-I It looks like the `walk(num_to_str)` is what you were talking about, right? Can we comment it out and test performance?

**Comment by bluddy** [comment]
@ZyX-I Sorry I guess I misunderstood and you already did this, but is it possible to get the time without the conversion component? As if you were doing the conversion with pure C?

**Comment by bluz71** [comment]
Slow Macbook (Core-M processor) result when converting from `vim.eval('a:items')` to `Nvim.eval(vim, "a:items")` against 60,000 file Linux source tree (an extreme large tree).

CtrlP + cpsm goes from unusable slow to usable slow. This is a big change. I would like more speed, but this is a massive change for the better.

**Comment by bluz71** [comment]
Question, how to use `vim.eval` and `Nvim.eval` in the same Python Vim module whilst being compatible with both Vim and Neovim?

Vim does not understand `Nvim.eval(vim, largeList)`, obviously.

Ideally I would like something like:

```
if Neovim
  from neovim import Nvim
  items = Nvim.eval(vim, "a:items")
else
  import vim
  items = vim.eval('a:items')
end
```

But, the Python parser will still need to understand Nvim module which doesn't exist in Vim 8 land.

**Comment by bluddy** [comment]
@bluz71 You can try importing neovim and catch the ImportError exception, at which point you'll know.

**Comment by Shougo** [comment]
```python
from importlib import find_loader

if find_loader('neovim'):
 from neovim import Nvim
  items = Nvim.eval(vim, "a:items")
else
  import vim
  items = vim.eval('a:items')
```

**Comment by bluz71** [comment]
I wonder how Vim plugin authors will react to PR contained `if neovim` paths such as noted by @Shougo's above?

It will lead to uglier code. Any ideas for more elegance?

**Comment by tjdevries** [comment]
You could do something like this:

```python
# File: $ROOT/plugin_repo/vim_eval.py

from importlib import find_loader

if find_loader('neovim'):
  from neovim import Nvim
  vim_eval = Nvim.eval
else
  import vim
  vim_eval = vim.eval

```

Then in some other file you could do:

```python
from plugin_repo.vim_eval import vim_eval

# now just use it below
```

You could also try giving the `neovim` or `vim` object another attribute and have that be inside of the new attribute.

**Comment by nixprime** [comment]
Python 2.7's importlib doesn't have (or at least isn't documented as having) find_loader. Checking for ImportError is insufficient because the neovim module is available outside of Neovim:

```
$ python3
Python 3.5.2 (default, Sep 14 2017, 22:51:06) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import neovim
>>>
```

However, vim.eval("has('nvim')") works fine. I've pushed an update to cpsm.

**Comment by bluz71** [comment]
Wow, @nixprime has been keeping an eye on this epic discussion.

Many thanks @nixprime! *cpsm* is a great plugin, I blogged about it [recently](https://bluz71.github.io/2017/10/26/turbocharge-the-ctrlp-vim-plugin.html). Hopefully a few more folks are now using this excellent plugin.

I'll update *cpsm* and test out your update.

**Comment by bluz71** [comment]
@nixprime, very good. Not Vim 8 fast, due to Neovim architecture, but a massive improvement when doing *CtrlP* in a huge source tree as compared to before. Huge thanks.

About twice as fast as before feels about right, still a bit laggy with huge trees, but now **much** more usable. The numbers provided by @ZyX-I [here](https://github.com/neovim/neovim/issues/7063#issuecomment-340588826) seem spot on.

**Comment by ZyX-I** [comment]
@bfredl has already pointed out an existence of `vim.api` which I did not know about. So just stop trying using `Nvim` class, calling a method from the other class with its descendant as `self` is not a thing you must ever do “in production”, or is guaranteed to work; I used it as a quick hack to test things.

**Comment by ZyX-I** [comment]
Also Vim does not have `vim.api`, so you may check for AttributeError.

**Comment by bfredl** [comment]
Yes, the canonical check would be
```
import vim
try:
    vim_eval = vim.api.eval
except AttributeError:
    vim_eval = vim.eval
```

**Comment by nixprime** [comment]
Where is vim.api documented? I don't see it in Neovim's if_pyth.txt.

**Comment by justinmk** [comment]
@nixprime The Nvim API is documented at `:help api`. The python-client `vim.api` is just a wrapper around that.

**Comment by bfredl** [comment]
It maybe still should be mentioned in if_pyth.txt, not everyone reads python-client readme and we should make it more easy to discover.

**Comment by Asheq** [comment]
I'm also experiencing a lag (about 1 to 2 seconds) when loading a file for the first time. Here is an excerpt from a profile I captured:

```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    1   1.100826   0.612652  <SNR>53_InitializePythonBuiltin()
    1   0.599849   0.599533  provider#pythonx#Require()
    1              0.487010  <SNR>98_check_interpreter()
    2   0.614059   0.014168  provider#python#Call()
    3   0.007806   0.007125  <SNR>11_LoadFTPlugin()
    1   1.126939   0.004908  <SNR>88_open_selected()
    3   0.010472   0.002834  <SNR>14_SynSet()
    1   0.002909   0.002001  <SNR>53_UseConfigFiles_Python_Builtin()
    1   0.001599   0.001583  <SNR>53_FindPythonFiles()
    4   0.001606   0.001579  <SNR>88_globlist()
    3   0.001547   0.001323  <SNR>12_LoadIndent()
    3   0.004870   0.000809  fugitive#detect()
    3              0.000694  <SNR>60_define_commands()
    1   0.000908   0.000622  <SNR>53_ApplyConfig()
    2   0.002105   0.000499  <SNR>88_list_dir()
   18   0.000705   0.000445  <SNR>60_buffer_spec()
    5              0.000444  <SNR>73_Highlight_Matching_Pair()
    2   0.000771   0.000419  fugitive#extract_git_dir()
    2   0.024237   0.000328  <SNR>88_open_dir()
    3   0.000614   0.000305  <SNR>60_buffer_type()
```

As you can see, the issue is Python-related. I disabled plugins until I found the offending plugin: [editorconfig-vim](https://github.com/editorconfig/editorconfig-vim). It uses Python. I replaced the plugin with [editorconfig-vim](https://github.com/sgur/vim-editorconfig), a variant that does not use Python (vimscript only) which resulted in the lag being removed when loading that first file.

_(Apologies if this isn't strictly related. I can move to a separate issue if that is the case)._

**Comment by jamessan** [comment]
@Asheq Your scenario is more likely related to _finding_ the Python interpreter to use, rather than the performance of the plugin itself -- `:help g:python_host_prog`.

**Comment by abarker** [comment]
I'm having a similar performance [issue with my plugin Cyfolds](https://github.com/abarker/cyfolds/issues/1) which computes Vim folds for Python (and is written in Python/Cython).  On Vim the code is quite fast, as most of the calculations are done in Cython in one pass.  The computed values are cached so the Vim buffer is accessed only once per fold update.  On Neovim the performance is relatively slow.

I rewrote the code to use the `vim.api.eval` method described in earlier comments, but that doesn't seem to make an appreciable difference.  Here is the code segment:
```
try: # See if in Neovim.
    vim_eval = vim.api.eval
    vim_command = vim.api.command
    def get_current_buf():
        return vim.api.get_current_buf()
except AttributeError: # No, in regular Vim.
    vim_eval = vim.eval
    vim_command = vim.command
    def get_current_buf():
        return vim.current.buffer
```

* `nvim --version`: NVIM v0.2.2
* Vim version: VIM - Vi IMproved 8.0
* Operating system/version: Ubuntu 18.04
* Terminal name/version: Gnome Terminal

These are the top few numbers I'm getting from opening a file of ~1600 lines in Vim/Neovim.  The function `GetPythonFoldViaCython()` is just calling `python3 call_get_foldlevel()` and returning a value which was set as a global variable in `call_get_foldlevel()`.

*nvim results:*
```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
 1643   1.649298             provider#python3#Call()
 1628   1.480020             GetPythonFoldViaCython()
    1   0.103830   0.022131  <SNR>54_OnFileTypeSet()
    1   0.044677   0.003110  <SNR>7_LoadFTPlugin()
    1   0.028734             UltiSnips#SnippetsInCurrentScope()
```

*vim results:*
```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    1   0.089110   0.010934  <SNR>51_OnFileTypeSet()
 1628   0.077059             GetPythonFoldViaCython()
    1   0.069215   0.033286  <SNR>12_LoadFTPlugin()
    1   0.025583             UltiSnips#SnippetsInCurrentScope()
    1   0.011860             pymode#breakpoint#init()
```

I'm not really familiar with the Neovim plugin API, so any suggestions would be appreciated.

**Comment by justinmk** [comment]
> On Vim the code is quite fast, as most of the calculations are done in Cython in one pass

But if you're setting the function as 'foldexpr' then that doesn't mean much because it gets called in proportion to the number of lines in the buffer. And your profile results make that pretty clear, it's called 1628 times which results in 1628 RPC calls. 

> ... `:python3 call_get_foldlevel()`

Each `:python3` invocation calls `provider#python3#Call()` which performs RPC to the python process.

If you want to use python and still have it be fast, you would need to do calculations python-side then send the results as a batch, and keep the batched results Vimscript-side, so that `GetPythonFoldViaCython()` only consults the batch, and does not keep calling python.

Meanwhile, Lua is faster that cython, and Vim9 is deprecating `if_pyth` ...

**Comment by bluz71** [comment]
> Meanwhile, LuaJit is faster that cython, and Vim9 is deprecating `if_pyth` ...

This sentence is especially important. Python-based plugins will likely incur the same RPC slowness in Vim 9; python interface will be phased out replaced by JSON-based RPC (which sounds similar to how Neovim does things). The Python free lunch will end at some point.

**Comment by abarker** [comment]
@justinmk Thanks for the help.  I didn't know that the `python3` calls would incur so much overhead just to pass a few characters back from a Python array.  It was an easy enough change in `GetPythonFoldViaCython` to only call  `call_get_foldlevel()` on the first line and set the values in a Vimscript array to return for later lines.  

I'm now getting numbers like this:
```
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
   16   0.460993             provider#python3#Call()
 1628   0.249488             GetPythonFoldViaCython()
    1   0.120687   0.024079  <SNR>54_OnFileTypeSet()
    1   0.037353   0.001752  <SNR>7_LoadFTPlugin()
    1   0.033571             UltiSnips#SnippetsInCurrentScope()
```
Definitely an improvement.  Not Vim 8 fast, but it should be usable.  Even the Vim 8 code runs a little faster now.

**Comment by justinmk** [comment]
@abarker Nice! I'm going to lock this thread so that your case study does not get lost in further comments, since it is a perfect illustration of what is discussed here.
